# Python Workshop

## NumPy

### Motivation

NumPy(Numerical Python)NumPy 旨在提供一种比传统 Python 列表快 50 倍的数组对象。它允许你在Python中进行**向量和矩阵计算**，并且由于许多底层函数实际上是用C编写的，因此你可以体验在原生Python中永远无法体验到的速度。

NumPy 中的数组对象称为 `ndarray`，它提供了许多支持函数，使操作变得非常简便。

数组在数据科学领域应用广泛，因为速度和资源利用率在数据科学中至关重要。

### Array

In [572]:
import numpy as np # Now the NumPy package can be referred to as np instead of numpy

# Create 1D Array
a = np.array([0, 1, 2, 3, 4])
b = np.array((0, 1, 2, 3, 4))
c = np.arange(5) # numpy.arange([start, ]stop, [step, ]), start默认为0，step默认为1
d = np.linspace(0, 2*np.pi, 5) # 创建范围为[0,2pi]，共5个元素的数组
print(a)
print(type(a))
print(b)
print(c)
print(d)



[0 1 2 3 4]
<class 'numpy.ndarray'>
[0 1 2 3 4]
[0 1 2 3 4]
[0.         1.57079633 3.14159265 4.71238898 6.28318531]


* type(): 这个内置的 Python 函数可以告诉我们传递给它的对象的类型。就像上面的代码一样，它显示 a 是 `numpy.ndarray` 类型。
  
要创建 `ndarray`，我们可以将列表、元组或任何类似数组的对象传递给 `array()` 方法，它将被转换为 `ndarray`。

In [573]:
# 0-D array
arr0 = np.array(42)
print(arr0)

# 1-D array
arr1 = np.array([1, 2, 3, 4, 5])
print(arr1)

# 2-D array
arr2 = np.array([[1, 2, 3], [4, 5, 6]])
print(arr2)

# check number of dimensions
print(arr0.ndim)
print(arr1.ndim)
print(arr2.ndim)

42
[1 2 3 4 5]
[[1 2 3]
 [4 5 6]]
0
1
2


### Higher Dimensional Arrays
创建数组时，可以使用 `ndmin` 参数定义维度数。

In [574]:
# Create an array with 5 dimensions
arr = np.array([1, 2, 3, 4], ndmin = 5)

print(arr)
print('number of dimensions: ', arr.ndim)

[[[[[1 2 3 4]]]]]
number of dimensions:  5


### NumPy Array Indexing
NumPy 数组中的索引从 0 开始，这意味着第一个元素的索引为 0，第二个元素的索引为 1，依此类推。

In [575]:
# Access 1-D array
arr = np.array([1, 2, 3, 4])
print(arr[0])
print(arr[1] + arr[2])

# Access 2-D Arrays [(row-1),(col-1)]
arr = np.array([[1,2,3,4,5], [6,7,8,9,10]])
print('2nd element on 1st row: ', arr[0, 1])
print('5th element on 2nd row: ', arr[1, 4])

1
5
2nd element on 1st row:  2
5th element on 2nd row:  10


要访问 n 维数组，我们使用 `arr_name[(dim_1st - 1),(dim_2nd - 1),...,(dim_nth - 1)]`

### Negative Indexing
我们使用负索引从数组末尾访问数组。

In [576]:
arr = np.array([[1,2,3,4,5], [6,7,8,9,10]])

print('Last element from 2nd dim: ', arr[1, -1])

Last element from 2nd dim:  10


一些有用的创建函数

1. numpy.eye()是 NumPy 库中用于创建一个二维数组（矩阵），在指定的对角线上为1，其他位置为零

In [577]:
print(np.eye(3)) #单位矩阵
print(np.eye(3, 5))
print(np.eye(3, 5, k = 0))
print(np.eye(3, 5, k = 1))

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
[[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]]
[[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]]
[[0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]]


2. numpy.diag() 是 NumPy 库中用于创建或提取对角线数组的函数

In [578]:
print(np.diag([1, 2, 3]))#对角矩阵

print(np.diag([1, 2, 3], 1))

a = np.array([[1, 2], [3, 4]]) #返回对角的值
b = np.array([[1, 2, 3],[4, 5, 6]])
print(np.diag(a))
print(np.diag(b))

[[1 0 0]
 [0 2 0]
 [0 0 3]]
[[0 1 0 0]
 [0 0 2 0]
 [0 0 0 3]
 [0 0 0 0]]
[1 4]
[1 5]


3. numpy.zeros() 用于初始化所有元素为0，numpy.ones() 用于初始化所有元素为1

In [579]:
print(np.zeros((2,3)))
print(np.zeros((2, 3, 2)))
print(np.ones((2, 3)))
print(np.ones((2, 3, 2)))

[[0. 0. 0.]
 [0. 0. 0.]]
[[[0. 0.]
  [0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]
  [0. 0.]]]
[[1. 1. 1.]
 [1. 1. 1.]]
[[[1. 1.]
  [1. 1.]
  [1. 1.]]

 [[1. 1.]
  [1. 1.]
  [1. 1.]]]


4. 创建随机矩阵

`numpy.random`模块实现了**伪随机数**生成器（PRNG 或简称 RNG），能够从各种概率分布中抽取样本。通常，用户会使用`default_rng`创建一个`Generator`实例，并调用其各种方法来获取不同分布的样本。

In [580]:
from numpy.random import default_rng
rng1 = np.random.default_rng(42)
print(rng1.random((2,3)))

rng2 = np.random.default_rng(42) 
print(rng2.random((2,3,2)))

# 42为种子值，可以用secrets.randbits(128)来获取任意 128 位整数作为种子值


[[0.77395605 0.43887844 0.85859792]
 [0.69736803 0.09417735 0.97562235]]
[[[0.77395605 0.43887844]
  [0.85859792 0.69736803]
  [0.09417735 0.97562235]]

 [[0.7611397  0.78606431]
  [0.12811363 0.45038594]
  [0.37079802 0.92676499]]]


### Slicing

  Python中的切片意味着将元素从一个给定索引取出到另一个给定索引

  切片的语法

  ```python

  print(array[start:end])

  print(array[start:end:step])
  ```

  start默认为0，end默认为array的长度，step默认为1。**Caution**：start会包括在内而end不会被包括在内

In [581]:
arr = np.array([1, 2, 3, 4, 5, 6, 7])

print(arr[1:5])

print(arr[-3:]) # [end-3, end]

print(arr[-3:-1]) # array[end-3, end-1]

print(arr[1:5:2]) # Caution: [index1,index5), step = 2


[2 3 4 5]
[5 6 7]
[5 6]
[2 4]


  **Quiz**: 输出二维数组的
1. 奇数行偶数列元素
2. 第二行
3. 第三列
4. 第一行的第二第四个元素
5. 第二第四行的第一第三个元素(共四个数)

  **Answer**

In [582]:
a = np.array(
    [[11, 12, 13, 14, 15],
    [16, 17, 18, 19, 20],
    [21, 22, 23, 24, 25],
    [26, 27, 28 ,29, 30],
    [31, 32, 33, 34, 35]])
print(a[::2,1::2])
print()
print(a[1, :])
print()
print(a[:, 2])
print()
print(a[0,[1,3]])
print()
print(a[[1,1,3,3],[0,2,0,2]])


[[12 14]
 [22 24]
 [32 34]]

[16 17 18 19 20]

[13 18 23 28 33]

[12 14]

[16 18 26 28]


### Array Properties

In [583]:
a = np.array([
            [11, 12, 13, 14, 15],
            [16, 17, 18, 19, 20],
            [21, 22, 23, 24, 25],
            [26, 27, 28 ,29, 30],
            [31, 32, 33, 34, 35]])

print(type(a)) # >>><class 'numpy.ndarray'>
print(a.dtype) # >>>int64
print(a.size) # >>>25
print(a.shape) # >>>(5, 5)
print(a.itemsize) # >>>8
print(a.ndim) # >>>2
print(a.nbytes) # >>>200


<class 'numpy.ndarray'>
int64
25
(5, 5)
8
2
200


In [584]:
# Creating Arrays With a Defined Data Type
arr = np.array([1, 2, 3, 4], dtype='S')
print()
print(arr)
print(arr.dtype)
print()

# Converting Data Type on Existing Arrays
arr = np.array([1.1, 0, 3])
newarr = arr.astype(bool)
print(newarr)
print(newarr.dtype)


[b'1' b'2' b'3' b'4']
|S1

[ True False  True]
bool



如果给定的类型中元素无法进行类型转换，那么 NumPy 将会抛出一个**ValueError**异常。


- `a.shape`  
  数组的形状是它有多少行和列，上面的数组有5行和5列，所以它的形状是(5，5)。  
- `a.itemsize`  
  itemsize属性是每个项占用的字节数。这个数组的数据类型是int 64，一个int 64中有64位，一个字节中有8位，除以64除以8，你就可以得到它占用了多少字节，在本例中是8。  
- `a.ndim`  
  ndim 属性是数组的维数。这个有2个。例如，向量只有1。  
- `a.nbytes`  
  nbytes 属性是数组中的所有数据消耗掉的字节数。你应该注意到，这并不计算数组的开销，因此数组占用的实际空间将稍微大一点。

### Caculation

术语“broadcast”描述了 NumPy 在算术运算期间如何处理不同形状的数组。受到某些限制，较小的数组会“broadcast”到较大的数组上，以便它们具有兼容的形状。broadcast提供了一种向量化数组操作的方法，以便循环发生在 C 而不是 Python 中。它无需制作不必要的数据副本即可完成此操作，并且通常会实现高效的算法实现。然而，在某些情况下，broadcast并不是一个好主意，因为它会导致内存使用效率低下，从而减慢计算速度。

In [585]:
a = np.array([1.0, 2.0, 3.0])
b = np.array([2.0, 2.0, 2.0])
print(a * b)

a = np.array([1.0, 2.0, 3.0])
b = 2.0
print(a * b)

print(a+b)

[2. 4. 6.]
[2. 4. 6.]
[3. 4. 5.]


In [586]:
from numpy import array, argmin, sqrt, sum
observation = array([111.0, 188.0])
codes = array([
               [102.0, 203.0],
               [132.0, 193.0],
               [45.0, 155.0],
               [57.0, 173.0]])
diff = codes - observation    # the broadcast happens here
print(diff)

[[ -9.  15.]
 [ 21.   5.]
 [-66. -33.]
 [-54. -15.]]


  想要更深入地了解[参见](https://numpy.org/doc/stable/user/basics.broadcasting.html#broadcastable-arrays)

  - 点乘与叉乘

In [587]:
# 点乘
a = np.array([1,0,0])
b = np.array([0,1,0])
print(np.dot(a,b))

0


In [588]:
# 叉乘
a = np.array([2,0,0])
b = np.array([2,2,0])
print(np.cross(a,b)) 

a = np.array([1,0,0])
b = np.array([0,1,0])
print(np.cross(a,b))
print(np.cross(b,a))

[0 0 4]
[0 0 1]
[ 0  0 -1]


- 逆矩阵和转置

In [589]:
# Get Inverse and Transpose

a=np.matrix([[1, 2, 3],[4, 5, 6]])

print(a)

print(a.I)

print(a.T)

[[1 2 3]
 [4 5 6]]
[[-0.94444444  0.44444444]
 [-0.11111111  0.11111111]
 [ 0.72222222 -0.22222222]]
[[1 4]
 [2 5]
 [3 6]]


[参见](https://numpy.org/doc/stable/reference/generated/numpy.matrix.html)

- 求和，最大最小值，累积求和（从前往后不断加）

In [590]:
a = np.arange(10)
print(a)
print(a.sum()) # >>>45
print(a.min()) # >>>0
print(a.max()) # >>>9
print(a.cumsum()) # >>>[ 0  1  3  6 10 15 21 28 36 45]

[0 1 2 3 4 5 6 7 8 9]
45
0
9
[ 0  1  3  6 10 15 21 28 36 45]


线性代数

In [591]:
from numpy import linalg as LA
matrix = np.diag((1, 2, 3))
print(matrix)
print()
eigenvalues, eigenvectors = LA.eig(matrix)
print(eigenvalues) # 特征值
print(eigenvectors) # 特征向量矩阵(按'列'存储)

[[1 0 0]
 [0 2 0]
 [0 0 3]]

[1. 2. 3.]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


[参见](https://numpy.org/doc/stable/reference/routines.linalg.html)

### 复制、拼接或修改现有的数组

1. 复制 `view` and `copy`

    view和copy可以类比为引用和拷贝，对于view的修改会直接作用到原始数据上，而copy则不会。
- copy  
copy拥有数据，对copy所做的任何更改都不会影响原始数组，而对原始数组所做的任何更改也不会影响copy。

In [592]:
arr = np.array([1, 2, 3, 4, 5])
x = arr.copy()
arr[0] = 42

print(arr)
print(x)

[42  2  3  4  5]
[1 2 3 4 5]


- view  
view并不拥有数据，对view所做的任何修改都会影响原始数组，而对原始数组所做的任何修改也会影响view。(和切片类似)


In [593]:
arr = np.array([1, 2, 3, 4, 5])
x = arr.view()
arr[0] = 42

print(arr)
print(x)

[42  2  3  4  5]
[42  2  3  4  5]


In [594]:
a = np.array([1, 2, 3, 4, 5, 6])
b = a[:2]
b +=1
print('a =', a, '; b =', b)

a = [2 3 3 4 5 6] ; b = [2 3]


如何判断是否是副本：  
- 经验方法：低维索引(切片/内存连续)通常会产生view而高维(内存非连续)通常为copy
- 代码测试：
    ```python
    x.base is None
    ```

**quiz**: What is the output of the following code?
``` python
a = np.arange(10)
b = a[2:6]
b = b * 2
print(a)
```
A. [2 3 4 5]
B. [0 1 2 3 4 5 6 7 8 9]
C. [ 4  6  8 10]
D. [ 0  1  2  3  4  5  6  7  8  9 10]


2. matrix的拼接

In [595]:
A = np.ones((2, 2))
B = np.eye(2, 2)
C = np.zeros((2, 2))
D = np.diag((-3, -4))
np.block([[A, B], [C, D]])

array([[ 1.,  1.,  1.,  0.],
       [ 1.,  1.,  0.,  1.],
       [ 0.,  0., -3.,  0.],
       [ 0.,  0.,  0., -4.]])

3. Reshape

    将一个matrix保持内容不变地变换形状,但要求reshape前后两个矩阵所含元素数量相同，否则会报错！

In [596]:
arr = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12])

newarr = arr.reshape(4, 3) # change to a 4*3 matrix

print(newarr)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]


`reshape()`允许有一个“未知”维度。这意味着在 reshape 方法中，您不必为其中一个维度指定确切的数值。将 -1 作为值传递，NumPy 将为您计算该数字。

In [597]:
arr = np.array([1, 2, 3, 4, 5, 6, 7, 8])

newarr = arr.reshape(2, 2, -1)

print(newarr)

[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]


将高维变成一维

In [598]:
arr = np.array([[1, 2, 3], [4, 5, 6]])

newarr = arr.reshape(-1)
newarr2 = arr.flatten() # 后面这两个只能把矩阵变为1-D数组
newarr3 = arr.ravel()

print(newarr)
print(newarr2)
print(newarr3)

[1 2 3 4 5 6]
[1 2 3 4 5 6]
[1 2 3 4 5 6]


**Quiz**: 将矩阵[1,2,3]改写成[[1,2,3],[4,5,6],[7,8,9]]再将矩阵转化为一维数组。使用broadcast,拼接,reshape等技巧。  
**Answer**:

In [599]:
# opt 1:
a = np.array([1, 2, 3])
b = a.copy() + 3 # 不可以用view！
c = a.copy() + 6

new_arr = np.block([[a],[b],[c]])
print(new_arr)

new_arr2 = new_arr.reshape(-1)
print(new_arr2)
print()

# opt 2 (use broadcast!):
a = np.array([1, 2, 3])
row_offset = np.array([0, 3, 6]).reshape(3,1)

new_arr = a + row_offset
print(new_arr)

new_arr2 = new_arr.ravel()
print(new_arr2)


[[1 2 3]
 [4 5 6]
 [7 8 9]]
[1 2 3 4 5 6 7 8 9]

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[1 2 3 4 5 6 7 8 9]


4. 布尔索引（Boolean Indexing）  

    用 True / False 作为“筛选器”
- 满足条件 → True → 被选中
- 不满足 → False → 被过滤

In [600]:
a = np.array([1,2,3,4,5,6]) # Masking
print(a[a<4])

[1 2 3]


In [601]:
a = np.arange(0, 100, 10)
b = a[:5] # from 0th element to 4th element
c = a[a >= 50]

print(a) # >>>[ 0 10 20 30 40 50 60 70 80 90]
print(b) # >>>[ 0 10 20 30 40]
print(c) # >>>[50 60 70 80 90]

[ 0 10 20 30 40 50 60 70 80 90]
[ 0 10 20 30 40]
[50 60 70 80 90]


In [602]:
# Where：返回值是满足条件的索引位置,而且返回值是一个 tuple!!(非常关键)
a = np.arange(0, 100, 10)
b = np.where(a < 50) 
c = np.where(a >= 50)[0]
print(a)
print(b) # >>>(array([0, 1, 2, 3, 4]),) 是一个长度为1的tuple
print(c) # >>>[5 6 7 8 9] 是ndarray

[ 0 10 20 30 40 50 60 70 80 90]
(array([0, 1, 2, 3, 4]),)
[5 6 7 8 9]


### Vectorization

vectorize把“只能处理单个元素的函数”
变成“可以作用在数组上的函数”。  
vectorize提供该功能主要是为了方便，而不是为了性能。该实现本质上是一个 for 循环。

In [603]:
def myfunc(a, b):
    "Return a-b if a>b, otherwise return a+b"
    if a > b:
        return a - b
    else:
        return a + b

vfunc = np.vectorize(myfunc)
print(vfunc([1, 2, 3, 4], 2))

# equivalence code
a = [1,2,3,4]
print(np.array([myfunc(x,2) for x in a]))


[3 4 1 2]
[3 4 1 2]


Answer of the quiz:  
1.**quiz**: What is the output of the following code?
``` python
a = np.arange(10)
b = a[2:6]
b = b * 2
print(a)
```
A. [2 3 4 5]  
**B. [0 1 2 3 4 5 6 7 8 9]**  
C. [ 4  6  8 10]  
D. [ 0  1  2  3  4  5  6  7  8  9 10]


## Pandas
`Pandas`使我们能够处理大量数据，尤其适合处理结构化数据（如表格、时间序列等, 并依据统计学原理得出结论。
`Pandas`能够清理混乱的数据集，并使其变得易于阅读且具有相关性。


### Series 
Series是一个一维数组，可以存储任何类型的数据。

Labels:  
如果没有特别指定，则值将使用其索引号进行标记。第一个值的索引为 0，第二个值的索引为 1，依此类推。

In [604]:
# Create series
import pandas as pd

a = [1, 7, 2]
myvar = pd.Series(a) 

print(myvar)
print(myvar[0])

0    1
1    7
2    2
dtype: int64
1


使用 `index` 参数，也可以自定义标签名称（例如“x”、“y”、“z”……）。

In [605]:
a = [1, 7, 2]
myvar = pd.Series(a, index = ["x", "y", "z"])

print(myvar)
print(myvar["y"])

x    1
y    7
z    2
dtype: int64
7


创建 Series 时，也可以使用键值对对象(字典)。

**Note**：字典的键将成为标签。

In [606]:
calories = {"day1": 420, "day2": 380, "day3": 390}

# select only some of the items by using index argument
myvar = pd.Series(calories, index = ["day1", "day2"])

print(myvar)

day1    420
day2    380
dtype: int64


### DataFrames  
A Pandas DataFrame是一种二维数据结构，类似于二维数组或具有行和列的表格。  
**Note**: Series is like a column, a DataFrame is the whole table.

1. 创建一个Dataframe

In [607]:
data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)

print(df) 

   calories  duration
0       420        50
1       380        40
2       390        45


2. 定位行/列  
Pandas 使用 `loc` 属性返回一个或多个指定的行。

In [608]:
# Return row 0, refer to the row index
print(df.loc[0])
print()
print(df.loc[[0,1]])

calories    420
duration     50
Name: 0, dtype: int64

   calories  duration
0       420        50
1       380        40


3. 命名索引

In [609]:
newdf = pd.DataFrame(data, index = ["day1", "day2", "day3"])
print(newdf)
print()
#refer to the named index:
print(newdf.loc[["day2","day1"]])

      calories  duration
day1       420        50
day2       380        40
day3       390        45

      calories  duration
day2       380        40
day1       420        50


4. 数据导入  
(1) CSV (Comma Separated Values)

In [610]:
df = pd.read_csv('data.csv')

print(df)
print(df.to_string()) # print the entire DataFrame

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

In [611]:
print(df.head(3)) # print the first 10 rows
print(df.tail()) # print the last 5 rows

   Duration          Date  Pulse  Maxpulse  Calories
0        60  '2020/12/01'    110       130     409.1
1        60  '2020/12/02'    117       145     479.0
2        60  '2020/12/03'    103       135     340.0
    Duration          Date  Pulse  Maxpulse  Calories
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0


(2) JSON  
大数据集通常是以 JSON 格式进行存储或提取的。  
JSON 是纯文本格式，但具有对象的结构.

In [612]:
df = pd.read_json('data.json')
print(df) 

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.4
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


### 数据清洗  
数据清洗是指修复数据集中的错误数据。  
错误数据可能包括：

- 空单元格

- 数据格式错误

- 数据错误

- 重复数据

1. **空单元格**

(1) 删除包含空单元格的行

默认情况下，`df.dropna()` 方法会返回一个新的 DataFrame，而不会修改原始 DataFrame。

如果需要修改原始 DataFrame，请使用 `df.dropna(inplace = True)`。

In [613]:
df = pd.read_csv('data.csv')
print(df.iloc[[18,22,28]])
new_df = df.dropna()
print(new_df.iloc[[18,22,28]])

    Duration          Date  Pulse  Maxpulse  Calories
18        45  '2020/12/18'     90       112       NaN
22        45           NaN    100       119     282.0
28        60  '2020/12/28'    103       132       NaN
    Duration          Date  Pulse  Maxpulse  Calories
19        60  '2020/12/19'    103       123     323.0
24        45  '2020/12/24'    105       132     246.0
31        60  '2020/12/31'     92       115     243.0


（2）替换空值

处理空单元格的另一种方法是插入新值。

In [614]:
# Replace all empty cells
df = pd.read_csv('data.csv')
df.fillna(130, inplace = True)
print(df.iloc[[18,22,28]])

    Duration          Date  Pulse  Maxpulse  Calories
18        45  '2020/12/18'     90       112     130.0
22        45           130    100       119     282.0
28        60  '2020/12/28'    103       132     130.0


In [615]:
# Replace only for specified columns
df.fillna({"Calories": 130}, inplace = True)
print(df.iloc[[18,22,28]])

# Replace using mean
df = pd.read_csv("data.csv")
x = df["Calories"].mean()
df.fillna({"Calories": x}, inplace = True)

# Replace using median
x = df["Calories"].median()
df.fillna({"Calories": x}, inplace = True)

# Replace using mode (the value that appears most frequently)
x = df["Calories"].mode()[0]
df.fillna({"Calories": x}, inplace = True)

    Duration          Date  Pulse  Maxpulse  Calories
18        45  '2020/12/18'     90       112     130.0
22        45           130    100       119     282.0
28        60  '2020/12/28'    103       132     130.0


2. **数据格式错误**

(1) 转换为正确格式

In [616]:
df = pd.read_csv('data.csv')
print(df.iloc[[21,22,26]])

df['Date'] = pd.to_datetime(df['Date'], format='mixed')
print(df.iloc[[21,22,26]])

    Duration          Date  Pulse  Maxpulse  Calories
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
26        60    2020/12/26    100       120     250.0
    Duration       Date  Pulse  Maxpulse  Calories
21        60 2020-12-21    108       131     364.2
22        45        NaT    100       119     282.0
26        60 2020-12-26    100       120     250.0


（2）删除整行

**Quiz**: Remember the command we have learned?  
df._______(subset=['Date'], inplace = True)

In [617]:
df.dropna(subset=['Date'], inplace = True)
print(df.iloc[[21,22,23]])

    Duration       Date  Pulse  Maxpulse  Calories
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0


3. 错误数据

e.g. 第 7 行的持续时间为 450，而其他所有行的持续时间都在 30 到 60 之间。

这不一定是错误的，但考虑到这是某人的锻炼记录数据集，我们可以得出结论：此人实际锻炼时间并未达到 450 分钟。

In [618]:
# (1) Replacing Values
print(df.iloc[[6,7]])
df.loc[7, 'Duration'] = 45
print(df.iloc[[6,7]])
# For wrong data for larger data sets
for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.loc[x, "Duration"] = 120

# (2) Removing Rows
for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.drop(x, inplace = True)

   Duration       Date  Pulse  Maxpulse  Calories
6        60 2020-12-07    110       136     374.0
7       450 2020-12-08    104       134     253.3
   Duration       Date  Pulse  Maxpulse  Calories
6        60 2020-12-07    110       136     374.0
7        45 2020-12-08    104       134     253.3


4. 重复项

(1) 发现重复项

In [619]:
print(df.iloc[[11,12]])
print()
print(df.duplicated())

    Duration       Date  Pulse  Maxpulse  Calories
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
dtype: bool


（2）删除重复项

In [620]:
df.drop_duplicates(inplace = True)
print(df.iloc[[11,12]])

    Duration       Date  Pulse  Maxpulse  Calories
11        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3


### 数据相关性

`corr()` 方法用于计算数据集中每一列之间的关系。

**注意**：`corr()` 方法会忽略 "not numeric"列。

In [621]:
df.corr()

,Duration,Date,Pulse,Maxpulse,Calories
Duration,1.000000,0.212084,-0.083417,-0.296585,0.350969
Date,0.212084,1.000000,-0.369328,-0.517827,-0.368101
Pulse,-0.083417,-0.369328,1.000000,0.261426,0.503243
Maxpulse,-0.296585,-0.517827,0.261426,1.000000,0.338515
Calories,0.350969,-0.368101,0.503243,0.338515,1.000000


如何解释结果？

- 数值介于 -1 到 1 之间。

- 1 表示存在一对一的关系（完全相关），对于此数据集，第一列中每个值增加时，另一个值也会增加。

- 0.9 也代表良好的相关性，如果增加一个值，另一个值很可能也会增加。

- -0.9 与 0.9 的相关性同样良好，但如果增加一个值，另一个值很可能会减少。

- 0.2 表示相关性很差，这意味着一个值增加并不意味着另一个值也会增加。

### 查看其他属性

In [622]:
print(pd.__version__) # Checking Pandas Version
print()
print(df.info()) # Print information about the data

2.3.3

<class 'pandas.core.frame.DataFrame'>
Index: 30 entries, 0 to 31
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   Duration  30 non-null     int64         
 1   Date      30 non-null     datetime64[ns]
 2   Pulse     30 non-null     int64         
 3   Maxpulse  30 non-null     int64         
 4   Calories  28 non-null     float64       
dtypes: datetime64[ns](1), float64(1), int64(3)
memory usage: 1.4 KB
None


### Reference
[2023 zzjc's part](https://github.com/UMJI-SSTIA/Python-wksp)

useful self-study website:  
[numpy](https://www.w3schools.com/python/numpy/numpy_array_reshape.asp)  
[panda](https://www.w3schools.com/python/pandas/default.asp)